### PR TITLE
allow changing database at runtime and add checking for initialisation

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**
Return error when SM used uninitialised

### CODE REVIEW

**Comment by paul-szczepanek-arm** [comment]
@pan- is that enough to keep pal consistent?

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 1]
```diff
@@ -92,26 +79,39 @@ ble_error_t GenericSecurityManager::init(
...
+            return BLE_ERROR_OPERATION_NOT_PERMITTED;
+        }
     }
 
+    ble_error_t result = init_database(db_path);
+    if (result != BLE_ERROR_NONE) {
+        return result;
+    }
+
+    result = init_database(db_path);
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: oops, will fix this

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 2]
```diff
@@ -750,8 +768,51 @@ ble_error_t GenericSecurityManager::oobReceived(
 // Helper functions
 //
 
+ble_error_t GenericSecurityManager::init_database(
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: this is just moved

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 3]
```diff
@@ -750,8 +768,51 @@ ble_error_t GenericSecurityManager::oobReceived(
...
+    if (!_db) {
+        return BLE_ERROR_NO_MEM;
+    }
+
+    _db->restore();
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::init_resolving_list() {
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: this is just moved

Comment (paul-szczepanek-arm) [reply features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: needs signature changed

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp (hunk 2 above)
Comment (pan-) [reply features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: That's perfect to move it into a function; it makes code more readable.

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 4]
```diff
@@ -92,26 +79,34 @@ ble_error_t GenericSecurityManager::init(
...
+            return BLE_ERROR_OPERATION_NOT_PERMITTED;
+        }
+    }
+
+    ble_error_t result = init_database(db_path);
+    if (result != BLE_ERROR_NONE) {
+        return result;
     }
 
+    init_resolving_list();
```
Comment (pan-) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: we should handle return from this function: `return init_resolving_list();` ?

**Pull request merged**